### PR TITLE
abstracted query "tying"

### PR DESCRIPTION
**Pull request by ethn**
It's increasingly evident that no single strategy ("join", "where exists", etc) will work everywhere on a large site; there has to be some capacity to optimize card queries.

This PR is a big step in that direction, because it makes it so that you can easily change strategies.

The API could use some further thought, and I need to spend some time organizing Query tests, but when I tweak the DEFAULT_FASTEN value by hand, all query_spec tests for all three values. 

Note that we could easily extend this same approach and add an option to break a single query into multiple ones.

### CODE REVIEW

File: card/lib/card/query/abstract_query/tie.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      #
+      # In concept, here's how the different strategies would tie table A to table B
+      # in SQL assuming A.id = B.a_id
+      #
+      # - :join  ...  FROM A JOIN B ON A.id = B.a_id
+      # - :exist ...  FROM A WHERE EXISTS (SELECT * FROM B WHERE A.id = B.a_id ...)
+      # - :in    ...  FROM A WHERE A.id IN (SELECT B.a_id FROM B WHERE ...)
+      #
+      # The different strategies will return the same values but the relative speed is
+      # context dependent.
```
Comment (ethn) [inline card/lib/card/query/abstract_query/tie.rb]: This explains the tying strategies.

**Comment by ethn** [comment]
For now, the default is set back to :join, so it's basically back to the old style queries for now, and then we can cherry pick slow ones to optimize.

File: card/lib/card/query/abstract_query/tie.rb (hunk 1 above)
Comment (xithan) [reply card/lib/card/query/abstract_query/tie.rb]: Where are the tests that prove that it really works like that?

Comment (ethn) [reply card/lib/card/query/abstract_query/tie.rb]: coming.

File: card/spec/lib/card/query/card_query/special_attributes_spec.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      it "it can handle *" do
+        @query = { match: "*all" }
+        is_expected.to include("*all")
+      end
+    end
+  end
+
+  describe "found_by" do
+    before do
+      Card::Auth.as_bot
```
Comment (xithan) [inline card/spec/lib/card/query/card_query/special_attributes_spec.rb]: I think doing it like this makes it persistent through all tests executed afterwards
Better use the `as_bot: true` option.

File: card/spec/lib/card/query/card_query/special_attributes_spec.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
+require_relative "../query_spec_helper"
+RSpec.describe Card::Query::CardQuery::SpecialAttributes do
+  include QuerySpecHelper
+
+  # TODO: add specs for: #name_match, #complete, #junction_complete
+
+  describe "match" do
+    it "reachs content and name via shortcut" do
+      @query = { match: "two" }
+      is_expected.to eq(cards_matching_two)
```
Comment (xithan) [inline card/spec/lib/card/query/card_query/special_attributes_spec.rb]: The style guide (which is disabled for theses files) dislikes instances variables and he has a point.
It obfuscates what the test is doing.

File: card/lib/card/query/abstract_query/tie.rb (hunk 1 above)
Comment (ethn) [reply card/lib/card/query/abstract_query/tie.rb]: (they are there now)

File: card/spec/lib/card/query/card_query/special_attributes_spec.rb (hunk 3 above)
Comment (ethn) [reply card/spec/lib/card/query/card_query/special_attributes_spec.rb]: I agree. This PR didn't introduce that, though. Can we go ahead and merge this?  I want to tackle the actual performance problems before getting into more spec clean up.

File: card/spec/lib/card/query/card_query/special_attributes_spec.rb (hunk 2 above)
Comment (ethn) [reply card/spec/lib/card/query/card_query/special_attributes_spec.rb]: for now I just used the as_bot block.  That way the queries are run with standard permissions.

File: card/spec/lib/card/query/card_query/special_attributes_spec.rb (hunk 3 above)
Comment (ethn) [reply card/spec/lib/card/query/card_query/special_attributes_spec.rb]: ok, I took care of it.  should be ready now.

**Pull request merged**